nixos/modules/services/networking/unbound.nix

```diff
@@ -6,41 +6,113 @@ let
+        default = {
+          directory = cfg.stateDir;
+          username = cfg.user;
+          chroot = cfg.stateDir;
+          pidfile = ''""'';
+          interface = [ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1");
+          access-control = [ "127.0.0.0/24 allow" ];
+          auto-trust-anchor-file = if cfg.enableRootTrustAnchor then rootTrustAnchorFile else null;
+        };
+        description = "server: clause configuration";
+      };
+      remote-control = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = {
+          control-enable = false;
+          control-interface = [ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1");
+          server-key-file = "${cfg.stateDir}/unbound_server.key";
+          server-cert-file = "${cfg.stateDir}/unbound_server.pem";
+          control-key-file = "${cfg.stateDir}/unbound_control.key";
+          control-cert-file = "${cfg.stateDir}/unbound_control.pem";
+        };
+        description = "remote-control: clause configuration";
+      };
+      stub-zone = mkOption {
+        type = with types; listOf (attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes))));
+        default = [ ];
+        description = "stub-zone: clause configuration";
+      };
+      forward-zone = mkOption {
+        type = with types; listOf (attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes))));
+        default = [ ];
+        example = literalExample ''
+          [
+            {
+              name = ".";
+              forward-addr = "1.1.1.1@853#cloudflare-dns.com";
+            }
+            {
+              name = "example.org.";
+              forward-addr = [
+                "1.1.1.1@853#cloudflare-dns.com"
+                "1.0.0.1@853#cloudflare-dns.com"
+              ];
+            }
+          ];
+        '';
+        description = "forward-zone: clauses configuration";
+      };
+      python = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = { };
+        description = "python: clause configuration";
+      };
+    };
+  };
+
+  yesOrNo = v: if v then "yes" else "no";
+
+  toOption = n: v: "${toString n}: ${v}";
+
+  toConf = n: v:
+    if builtins.isFloat v then (toOption n (toString v))
+    else if isInt v       then (toOption n (toString v))
+    else if isBool v      then (toOption n (yesOrNo v))
+    else if isString v    then (toOption n v)
+    else if isList v      then (concatMapStringsSep "\n  " (toConf n) v)
+    else abort "unbound.toConf: unexpected type (n = ${n})";
+
+  clauseConf = clause: conf: if conf == { } then "" else
+    (concatStringsSep "\n  " (
+      ["${clause}:"] ++ (
+        mapAttrsToList toConf (
+          filterAttrs (_: v: v != null) conf
+        )
+      )));
+
+  confFile = pkgs.writeText "unbound.conf" (concatStringsSep "\n" (flatten [
```

Comment:
Is this a custom file format, or can it fit into one of our existing `generator` implementation? If we don't have a `generator` implementation for this format... should we include one in `lib`?

Reply:
Not that I know of. Maybe someone with more experience can think of one